### ISSUE TITLE
Collect non-benchmark tests which use a physical device

### ISSUE THREAD

**Issue by keyonghan** [flutter/flutter#142070]
Most non-benchmark tests should be able to run without a physical device, and an emulator should be used instead to save limited DeviceLab resources and potentially improve the test execution time.

This bug tracks to collect the list.

/cc @matanlurey @jonahwilliams @jmagman

**Comment by matanlurey** [comment]
Thank you! Here is an easy one:

```yaml
  - name: Linux_pixel_7pro external_ui_integration_test
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: external_textures_integration_test

  - name: Mac_ios external_ui_integration_test_ios
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "ios", "mac"]
      task_name: external_textures_integration_test_ios
      ignore_flakiness: "true"
```

These don't need a physical device.

**Comment by jmagman** [comment]
`external_ui_integration_test_ios` is using the `dev/devicelab/integration_test` runners which check which devices are running, and pass `-d` to the tool, and fail if there are no devices tethered.  So this change isn't just to change the bot properties, it would involve swapping them to a different or more flexible test runner, like moving them into the tools tests.

Same for:
`channels_integration_test_ios`
`route_test_ios`
`flavors_test_ios`
`flavors_test_ios_xcode_debug`
`integration_test_test_ios`
`integration_ui_ios_driver`
`integration_ui_ios_driver_xcode_debug`
and others on the `devicelab/devicelab_drone` recipe.

**Comment by jonahwilliams** [comment]
Some more tests that can be emulators

```
  - name: Linux_pixel_7pro android_defines_test
    recipe: devicelab/devicelab_drone
    presubmit: true
    timeout: 60
    dimensions:
      kvm: "1"
      cores: "8"
      machine_type: "n1-standard-8"
    properties:
      contexts: >-
        [
          "android_virtual_device"
        ]
      device_type: "none"
      tags: >
        ["devicelab", "linux"]
      task_name: android_defines_test
      dependencies: >-
        [
          {"dependency": "android_virtual_device", "version": "34"}
        ]

  - name: Linux_pixel_7pro android_obfuscate_test
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: android_obfuscate_test

  - name: Linux_pixel_7pro android_semantics_integration_test
    recipe: devicelab/devicelab_drone
    bringup: true # Flaky: https://github.com/flutter/flutter/issues/124636
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: android_semantics_integration_test

  - name: Linux_pixel_7pro basic_material_app_android__compile
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: basic_material_app_android__compile
      
 - name: Linux_pixel_7pro channels_integration_test
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: channels_integration_test


  - name: Linux_pixel_7pro drive_perf_debug_warning
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: drive_perf_debug_warning

- name: Linux_pixel_7pro embedded_android_views_integration_test
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: embedded_android_views_integration_test

  - name: Linux_pixel_7pro external_ui_integration_test
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: external_ui_integration_test

  - name: Linux_pixel_7pro flavors_test
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: flavors_test

  - name: Linux_pixel_7pro flutter_gallery_android__compile
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: flutter_gallery_android__compile


  - name: Linux_pixel_7pro frame_policy_delay_test_android
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: frame_policy_delay_test_android

  - name: Linux_pixel_7pro integration_ui_driver
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: integration_ui_driver

  - name: Linux_pixel_7pro integration_ui_keyboard_resize
    bringup: true # Flaky https://github.com/flutter/flutter/issues/140645
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: integration_ui_keyboard_resize
      
        - name: Linux_pixel_7pro integration_ui_textfield
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: integration_ui_textfield

  - name: Linux_pixel_7pro native_assets_android
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: native_assets_android

  - name: Linux_pixel_7pro platform_channel_sample_test
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: platform_channel_sample_test

  - name: Linux_pixel_7pro platform_interaction_test
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: platform_interaction_test

  - name: Linux_pixel_7pro routing_test
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: routing_test

  - name: Linux_pixel_7pro service_extensions_test
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: service_extensions_test


  - name: Linux_pixel_7pro android_choreographer_do_frame_test
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: android_choreographer_do_frame_test


  - name: Mac_pixel_7pro entrypoint_dart_registrant
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "mac", "pixel", "7pro"]
      task_name: entrypoint_dart_registrant
    runIf:
      - dev/**
      - packages/flutter_tools/**
      - bin/**
      - .ci.yaml

  - name: Mac_pixel_7pro hello_world_android__compile
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "mac", "pixel", "7pro"]
      task_name: hello_world_android__compile


  - name: Mac_pixel_7pro integration_test_test
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "mac", "pixel", "7pro"]
      task_name: integration_test_test


  - name: Mac_pixel_7pro integration_ui_frame_number
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "mac", "pixel", "7pro"]
      task_name: integration_ui_frame_number


  - name: Mac_pixel_7pro native_assets_android
    recipe: devicelab/devicelab_drone
    presubmit: false
    bringup: true
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "mac", "pixel", "7pro"]
      task_name: native_assets_android

  - name: Mac_pixel_7pro run_debug_test_android
    recipe: devicelab/devicelab_drone
    presubmit: false
    runIf:
      - .ci.yaml
      - dev/**
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "mac", "pixel", "7pro"]
      task_name: run_debug_test_android

  - name: Mac_pixel_7pro run_release_test
    recipe: devicelab/devicelab_drone
    presubmit: false
    runIf:
      - .ci.yaml
      - dev/**
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "mac", "pixel", "7pro"]
      task_name: run_release_test

```


Not sure about this one (@yjbanov ) but it looks like this should not be a device test at all

```

  - name: Linux_pixel_7pro flutter_gallery_v2_chrome_run_test
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: flutter_gallery_v2_chrome_run_test
      
        - name: Linux_pixel_7pro linux_chrome_dev_mode
    recipe: devicelab/devicelab_drone
    presubmit: false
    timeout: 60
    properties:
      tags: >
        ["devicelab", "android", "linux", "pixel", "7pro"]
      task_name: linux_chrome_dev_mode
      ```

**Comment by jmagman** [comment]
I can work on getting some of the iOS ones to run on a simulator, hopefully not much more complicated than piping through a device override:
https://github.com/flutter/flutter/blob/90e4a12f2ab475185d9a3c9b164e2e26870ee643/dev/devicelab/bin/tasks/hot_mode_dev_cycle_ios_simulator.dart#L13-L19

**Comment by keyonghan** [comment]
**Tl;dr**: please confirm the test feasibility to switch to a VM or a VM with an emulator using this sheet: [go/flutter_devicelab_to_vm_list](go/flutter_devicelab_to_vm_list).

Thanks @jonahwilliams @jmagman and @matanlurey for the initial input! I have scanned through all our benchmarks and created the test list combining what you have provided (categorized based on test owners).

There targets that ones that are NOT uploading metrics to Skia (NOT benchmarks), but are running with a Device.
Please help identify/confirm those that can be switched to a VM or a VM with an emulator (based on test ownership) using this sheet, and I will update the status here after finalized accordingly). Feel free to loop other stakeholders who may have more context. Thanks!

- @camsim99
  - Mac_ios spell_check_test
- @christopherfujino
  - Mac_ios ios_content_validation_test
  - Mac_arm64_ios ios_content_validation_test
- @dcharkes
  - Linux_pixel_7pro native_assets_android
  - Windows_android native_assets_android 
  - Mac_ios native_assets_ios
- @dnfield
  - Linux_pixel_7pro integration_ui_frame_number
  - Mac_ios integration_ui_ios_frame_number
- @gaaclarke
  - Linux_pixel_7pro animated_advanced_blend_perf_opengles__timeline_summary
  - Linux_pixel_7pro animated_advanced_blend_perf__timeline_summary
  - Linux_pixel_7pro animated_blur_backdrop_filter_perf_opengles__timeline_summary
  - Linux_pixel_7pro entrypoint_dart_registrant
  - Linux_pixel_7pro hello_world_impeller
  - Mac_ios wide_gamut_ios
- @HansMuller
  - Linux_pixel_7pro android_semantics_integration_test (confirmed by @jonahwilliams [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1907056971))
- @hellohuanlin
  - Mac_ios native_platform_view_ui_tests_ios
- @jonahwilliams 
  - Linux_pixel_7pro animated_blur_backdrop_filter_perf__timeline_summary
  - Linux_pixel_7pro draw_vertices_perf_opengles__timeline_summary
  - Linux_pixel_7pro draw_vertices_perf__timeline_summary
- @reidbaker
  - Linux_pixel_7pro android_choreographer_do_frame_test  
- @stuartmorgan
  - Linux_android hybrid_android_views_integration_test 
  - Linux_pixel_7pro embedded_android_views_integration_test (confirmed by @jonahwilliams [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1907056971))
  - Linux_pixel_7pro platform_interaction_test
  - Mac_ios ios_defines_test
  - Mac_ios ios_platform_view_tests
  - Mac_ios platform_view_ios__start_up
- @vashworth
  - Mac_arm64_ios module_test_ios
  - Mac_arm64_ios flavors_test_ios_xcode_debug (confirmed by @jmagman [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1906995342))
  - Mac_ios channels_integration_test_ios (confirmed by @jmagman [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1906995342))
  - Mac_ios flavors_test_ios (confirmed by @jmagman [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1906995342))
  - Mac_ios integration_test_test_ios (confirmed by @jmagman [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1906995342))
  - Mac_ios integration_ui_ios_driver (confirmed by @jmagman [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1906995342))
  - Mac_ios integration_ui_ios_driver_xcode_debug (confirmed by @jmagman [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1906995342))
  - Mac_ios integration_ui_ios_keyboard_resize
  - Mac_ios integration_ui_ios_textfield
  - Mac_ios platform_channel_sample_test_ios
  - Mac_ios platform_channel_sample_test_swift
  - Mac_ios platform_interaction_test_ios
  - Mac_ios route_test_ios
- @zanderso
  - Linux_android devtools_profile_start_test  
  - Linux_android flutter_view__start_up 
  - Linux_android platform_view__start_up
  - Linux_pixel_7pro android_defines_test (confirmed by @jonahwilliams [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1907056971))
  - Linux_pixel_7pro android_obfuscate_test (confirmed by @jonahwilliams [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1907056971))
  - Linux_pixel_7pro basic_material_app_android__compile (confirmed by @jonahwilliams [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1907056971) but is a benchmark test. Please **DOUBLE** confirm.)
  - Linux_pixel_7pro channels_integration_test (confirmed by @jonahwilliams [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1907056971))
  - Linux_pixel_7pro drive_perf_debug_warning (confirmed by @jonahwilliams [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1907056971))
  - Linux_pixel_7pro external_ui_integration_test (confirmed by @jonahwilliams [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1907056971))
  - Linux_pixel_7pro flavors_test (confirmed by @jonahwilliams [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1907056971))
  - Linux_pixel_7pro flutter_gallery_android__compile (confirmed by @jonahwilliams [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1907056971) but is a benchmark test. Please **DOUBLE** confirm.)
  - Linux_pixel_7pro frame_policy_delay_test_android (confirmed by @jonahwilliams [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1907056971) but is a benchmark test. Please **DOUBLE** confirm.)
  - Linux_pixel_7pro hello_world_android__compile
  - Linux_pixel_7pro imagefiltered_transform_animation_perf__timeline_summary
  - Linux_pixel_7pro integration_test_test
  - Linux_pixel_7pro integration_ui_driver (confirmed by @jonahwilliams [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1907056971))
  - Linux_pixel_7pro integration_ui_keyboard_resize (confirmed by @jonahwilliams [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1907056971))
  - Linux_pixel_7pro integration_ui_textfield
  - Linux_pixel_7pro linux_chrome_dev_mode (mentioned by @jonahwilliams [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1907056971) but is a benchmark test. Please **DOUBLE** confirm.)
  - Linux_pixel_7pro platform_channel_sample_test
  - Linux_pixel_7pro routing_test
  - Linux_pixel_7pro run_debug_test_android
  - Linux_pixel_7pro run_release_test (confirmed by @jonahwilliams [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1907056971))
  - Linux_pixel_7pro service_extensions_test
  - Mac_ios external_ui_integration_test_ios (confirmed by @matanlurey [above](https://github.com/flutter/flutter/issues/142070#issuecomment-1906919709))
  - Mac_ios flutter_view_ios__start_up
  - Windows_android channels_integration_test_win
  - Windows_android flavors_test_win
- @yjbanov 
  - Linux_pixel_7pro flutter_gallery_v2_chrome_run_test

**Comment by jonahwilliams** [comment]
```
Linux_pixel_7pro animated_blur_backdrop_filter_perf__timeline_summary
Linux_pixel_7pro draw_vertices_perf_opengles__timeline_summary
Linux_pixel_7pro draw_vertices_perf__timeline_summary
```

These are definitely benchmarks.

```
basic_material_app_android__compile
```

This is a benchmark but it is reporting the compile times from the host, so the target device is irrelevant.

```
linux_chrome_dev_mode
```

This looks like a web benchmark that is running on the host. @yjbanov would need to confirm.


```
frame_policy_delay_test_android
```

ACK, this one is actually a benchmark, strike it from the list.

**Comment by dcharkes** [comment]
> @dcharkes
> * Linux_pixel_7pro native_assets_android
> * Windows_android native_assets_android
> * Mac_ios native_assets_ios

At least for ios the simulator needs less signing than the physical device, so to keep test coverage for the signing, we should keep running that one on device.

For the Android ones I'd be a little uncomfortable if we don't test on any physical device. There might be some differences between emulator and physical device, but nothing pops to mind. cc @reidbaker @stuartmorgan

**Comment by jonahwilliams** [comment]
The tool literally cannot tell the difference between Android emulators/simulators, beyond architecture (in fact, a number of flagship phones claim to be emulators in their adb config). As long as you can get an arm/arm64 emulator, there should not be any differences.

**Comment by dcharkes** [comment]
We did have this weird behavior difference between the emulator and physical device:

* https://github.com/dart-lang/sdk/issues/53249

But if we're struggling for capacity maybe we don't care about the minimal differences between device and emulator. @reidbaker Should probably make the call here.

**Comment by reidbaker** [comment]
At a high level I think that we should be able to run on real devices provided the tests are in post submit and real devices long-term are important for us to test on. 

I hear the arguments for why emulators are better, especially in presubmit where we have the ability to scale.

**Comment by jmagman** [comment]
Here's an example of a test moving out of devicelab runner into the tool integration tests: https://github.com/flutter/flutter/issues/70978 and https://github.com/flutter/flutter/pull/74685

**Comment by jmagman** [comment]
> Here's an example of a test moving out of devicelab runner into the tool integration tests: #70978 and #74685

To expand on this, some of the tests running in the devicelab or on devicelab recipe aren't doing anything on a target device, either physical or simulator.  They are building or otherwise validating apps on only a host machine.

The `runProjectTest` tests are good candidates to be moved into the tool integration tests.
android_obfuscate_test (currently using a Linux/Pixel devicelab bot)
engine_dependency_proxy_test (already running in a VM)
gradle_desugar_classes_test (VM)
gradle_plugin_bundle_test (VM)
gradle_plugin_fat_apk_test (VM)
gradle_plugin_light_apk_test (VM)
ios_content_validation_test (currently using a Mac/iOS devicelab bot) PR to move: https://github.com/flutter/flutter/pull/142272

**Comment by keyonghan** [comment]
> ```
> basic_material_app_android__compile
> ```
> 
> This is a benchmark but it is reporting the compile times from the host, so the target device is irrelevant.

Sounds a good fit for a host only benchmark (may not be suitable for a VM): https://github.com/flutter/flutter/issues/126090

**Comment by jmagman** [comment]
> Sounds a good fit for a host only benchmark (may not be suitable for a VM): #126090

https://github.com/flutter/flutter/pull/142280#discussion_r1468242596

> `hot_mode_dev_cycle` is the only test I'm aware of where we need a pinned physical device and pinned host hardware since the benchmark is measuring hot reload times, so time for the host to prep it and the device to apply it.
> 
> The `__compile` ones are hostonly benchmarks, they are gathering metrics for how long compilation/subsequent complication takes.  So they do not run on physical mobile devices, but should run on physical hosts pinned to consistent hardware.  Like I don't know why we have a `Mac_pixel_7pro hello_world_android__compile` and `Mac_arm64_android hello_world_android__compile` tests, they are doing the same and should be on `Mac_benchmark`.

**Comment by matanlurey** [comment]
@jmagman chatted with me offline in a group chat, and mentioned two bits for aspiring migrators:

> With our existing runners it's a lot of work to move out of `dev/devicelab/bin/tasks/`, the test essentially needs to be rewritten.  Here's an example where I moved one from that runner into the tools integration test runner https://github.com/flutter/flutter/pull/74685
>
> A more scalable way to do this would be to decouple `dev/devicelab/` from the actual devicelab and instead allow that runner to target a simulator, instead of the currently tethered physical device

/cc @jtmcdole

**Comment by jmagman** [comment]
> A more scalable way to do this would be to decouple `dev/devicelab/` from the actual devicelab and instead allow that runner to target a simulator, instead of the currently tethered physical device

(Similar to this https://github.com/flutter/flutter/issues/142070#issuecomment-1907135483)
